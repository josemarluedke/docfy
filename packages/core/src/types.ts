/* eslint-disable @typescript-eslint/no-explicit-any */

import { Node as MarkdownAST } from 'unist';
import { Processor, Plugin as UnifiedPlugin, Settings as UnifiedSettings } from 'unified';
import { VFile } from 'vfile';

export interface Heading {
  title: string;
  id: string;
  depth: number;
  headings?: Heading[];
}

export interface PageMetadata {
  url: string;
  relativeUrl: undefined | string;
  relativePath: string;
  editUrl: string;
  title: string;
  headings: Heading[];
  frontmatter: Record<string, unknown>;
  pluginData: Record<string, unknown>;
  parentLabel: undefined | string;
}

export interface PageContent {
  meta: PageMetadata;
  sourceConfig: SourceConfig;
  source: string;
  vFile: VFile;
  ast: MarkdownAST;
  markdown: string;
  rendered: string;
  demos?: PageContent[];
  pluginData: Record<string, unknown>;
}

interface ContextOptions
  extends Omit<Options, 'plugins' | 'remarkPlugins' | 'rehypePlugins' | 'tocMaxDepth'> {
  tocMaxDepth: number;
}

export interface StaticAssetDefinition {
  fromPath: string;
  toPath: string;
}

export interface Context {
  remark: Processor;
  rehype: Processor;
  pages: PageContent[];
  staticAssets: StaticAssetDefinition[];
  options: ContextOptions;
}

export interface NestedPageMetadata {
  name: string;
  label: string;
  pages: PageMetadata[];
  children: NestedPageMetadata[];
}

export interface DocfyResult {
  content: PageContent[];
  staticAssets: StaticAssetDefinition[];
  nestedPageMetadata: NestedPageMetadata;
}

export interface SourceConfig {
  /**
   * The absolute path to where the files are located.
   */
  root: string;

  /**
   * Match files using the patterns the shell uses, like stars and stuff. It
   * uses Glob package.
   */
  pattern: string;

  /**
   * Pattern to ignore.
   */
  ignore?: string[];

  /**
   * Option to change how urls are generated.
   *
   * There are two ways URLs can ge generated.
   *
   * 1. "auto": Uses the folder structure to inform how the urls are generated.
   *    For example, if you have the folowing files:
   *      - install.md
   *      - components/
   *        - button.md
   *        - card.md
   *    The urls would look like this, (assuming urlPrefix is set to "docs").
   *      - docs/install
   *      - docs/components/buttons
   *      - docs/components/card
   *
   * 2. "manual": It uses frontmatter information to inform "subcategory" and
   *    "category" of the file, ignoring the original file location.
   *    The url schema is as follows:
   *    {category}/{subcategory}/{file-name}
   *
   *    If no category or subcategory is specified, all files will be at the root
   *    level.
   *
   *    This options is perfect for documenting monorepo projects while keeping
   *    documentation next to implementation.
   */
  urlSchema?: 'auto' | 'manual';

  /**
   * Option to prefix the urls.
   *
   * For example:
   * "docs" will generate urls like "/docs/something"
   * "blog" will generate urls like "/blog/something"
   */
  urlPrefix?: string;

  /**
   * Option to suffix the urls.
   *
   * For example:
   * ".html" will generate urls like "/something.html"
   */
  urlSuffix?: string;

  /**
   * Overwrite repository config for this source.
   */
  repository?: RepositoryConfig;
}

interface RepositoryConfig {
  /**
   *
   * The url to the Git Repository
   * Example: https://github.com/josemarluedke/docfy
   */
  url: string;

  /**
   * Branch used to edit your markdown when clicking on "Edit this page" button.
   * @default "master"
   */
  editBranch?: string;
}

export interface PluginOptions {
  [key: string]: unknown;
}

export type PluginHandler<T = PluginOptions | undefined | null> = (
  ctx: Context,
  options: T
) => void;

export interface Plugin<T = PluginOptions | undefined | null> {
  runBefore?: PluginHandler<T>;
  runWithMdast?: PluginHandler<T>;
  runWithHast?: PluginHandler<T>;
  runAfter?: PluginHandler<T>;
}

export interface PluginWithOptions<T = PluginOptions> extends Plugin<T> {
  __options?: T;
}

export type PluginWithOptionsFunction<T = PluginOptions> = (options?: T) => PluginWithOptions<T>;

export type PluginList = (Plugin | PluginWithOptions<any> | PluginWithOptionsFunction<any>)[];

export interface Options {
  /**
   * A list of Docfy plugins.
   */
  plugins?: PluginList;

  /**
   * Additional remark plugins
   *
   * Example:
   *
   * ```js
   * const hbs = require('remark-hbs');
   * const autolinkHeadings = require('remark-autolink-headings');
   *
   * const remarkPlugins = [
   *   [
   *     autolinkHeadings,
   *     {
   *       behavior: 'wrap'
   *     }
   *   ],
   *   hbs
   * ];
   * ```
   */
  remarkPlugins?: ([UnifiedPlugin, UnifiedSettings] | UnifiedPlugin)[];

  /**
   * Additional rehype plugins
   *
   * Example:
   *
   * ```js
   * const rehypePrism = require('@mapbox/rehype-prism');
   *
   * const rehypePlugins = [rehypePrism];
   * ```
   */
  rehypePlugins?: ([UnifiedPlugin, UnifiedSettings] | UnifiedPlugin)[];

  /**
   * The max depth of headings
   * @default 6
   */
  tocMaxDepth?: number;

  /**
   * The repository config
   */
  repository?: RepositoryConfig;

  /**
   * Labels to be used while generating nestedPageMetadata
   */
  labels?: Record<string, string>;

  /**
   * The static asset path to be used in the their url.
   * Eg. staticAssetsPath: "/assets/docfy" and a static asset name as
   * "github-icon.png", the URL to be used for the image would be
   * "/assets/docfy/github-icon.png".
   *
   * @default "/assets/docfy"
   */
  staticAssetsPath?: string;
}

interface DocfyConfigSourceConfig extends Omit<SourceConfig, 'root'> {
  root?: string;
}

export interface DocfyConfig extends Options {
  sources: DocfyConfigSourceConfig[];
}
