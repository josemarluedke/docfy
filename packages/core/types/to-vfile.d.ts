declare module 'to-vfile' {
  import { VFile } from 'vfile';

  class ToVFile {
    static readSync(filepath: string): VFile;
  }

  export default ToVFile;
}
