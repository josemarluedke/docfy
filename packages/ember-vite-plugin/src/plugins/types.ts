import { Node } from 'unist';

interface Literal {
  value: string;
}

export interface CodeNode extends Node, Literal {
  type: 'code';
  lang?: string;
  meta?: string;
}

export interface DemoComponent {
  name: DemoComponentName;
  chunks: DemoComponentChunk[];
  description?: {
    title?: string;
    ast: Node;
    editUrl?: string;
  };
}

export interface DemoComponentName {
  dashCase: string;
  pascalCase: string;
}

export interface DemoComponentChunk {
  type: string;
  code: string;
  ext: string;
  snippet: Node;
}
