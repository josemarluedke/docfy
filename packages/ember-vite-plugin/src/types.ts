import type { Node } from 'unist';

interface Literal {
  value: string;
}
export interface CodeNode extends Node, Literal {
  type: 'code';
  lang?: string;
  meta?: string;
}
export interface ImportStatement {
  type: 'component' | 'helper' | 'modifier' | 'service' | 'other';
  name: string;
  path: string;
  isDefault: boolean;
  namedImports?: string[];
}

export interface InlineComponent {
  name: string;
  type: 'template-only' | 'class-based' | 'const-template';
  template: string;
  script?: string;
  imports?: ImportStatement[];
}

export interface GJSMetadata {
  imports: ImportStatement[];
  inlineComponents: InlineComponent[];
  additionalContent: string;
  templateContent: string; // The main HTML/HBS content
}

// Legacy demo component format from ember package
export interface LegacyDemoComponent {
  name: {
    pascalCase: string;
    dashCase: string;
  };
  chunks: Array<{
    ext: string;
    code: string;
    type: string;
  }>;
}

// New demo component format matching original ember implementation
export interface DemoComponentName {
  dashCase: string;
  pascalCase: string;
}

export interface DemoComponentChunk {
  type: string;
  code: string;
  ext: string;
  snippet: any; // AST node reference
}

export interface DemoComponent {
  name: DemoComponentName;
  chunks: DemoComponentChunk[];
  description?: {
    title?: string;
    ast: any; // AST node reference
    editUrl?: string;
  };
}

export interface PreviewTemplateComponent {
  name: string;
  template: string;
  ext: string; // File extension (hbs, gjs, gts, etc.)
  originalNode?: any; // Reference to AST node for replacement
}
