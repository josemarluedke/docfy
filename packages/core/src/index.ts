import glob from 'fast-glob';
import path from 'path';
import trough, { Through } from 'trough';
import toVfile from 'to-vfile';
import {
  PageContent,
  Context,
  Options,
  SourceConfig,
  PageMetadata,
  DocfyResult,
  Plugin,
  PluginList,
  PluginWithOptions,
  PluginWithOptionsFunction,
} from './types';
import { DEFAULT_IGNORE, generateAutoUrl, generateManualUrl, inferTitle } from './-private/utils';
import { createRehype, createRemark } from './-private/remark';
import {
  combineDemos,
  renderMarkdown,
  replaceInternalLinks,
  staticAssets,
  toc,
  uniquefyUrls,
  removeUnnecessaryIndex,
} from './plugins';
import { getRepoEditUrl } from './-private/repo-info';
import { transformToNestedPageMetadata } from './-private/nested-page-metadata';
import debugFactory from 'debug';
const debug = debugFactory('@docfy/core');

class Docfy {
  private pipeline: Through<Context>;
  private context: Context;
  private plugins: PluginList;

  constructor(options: Options = {}) {
    const { remarkPlugins, rehypePlugins, ...rest } = options;

    this.context = {
      remark: createRemark(remarkPlugins),
      rehype: createRehype(rehypePlugins),
      pages: [],
      staticAssets: [],
      options: {
        ...rest,
        tocMaxDepth: rest.tocMaxDepth || 6,
      },
    };

    const plugins: PluginList = [
      combineDemos,
      removeUnnecessaryIndex,
      uniquefyUrls,
      replaceInternalLinks,
      staticAssets,
    ];

    if (Array.isArray(options.plugins)) {
      plugins.push(...options.plugins);
    }

    // Make sure TOC and renderMarkdown plugins are the last ones
    plugins.push(toc, renderMarkdown);
    this.plugins = plugins;

    this.pipeline = trough<Context>()
      .use<SourceConfig[]>(this.initializePipeline.bind(this))
      .use(this.createPluginPipelineFor('runBefore'))
      .use(this.createPluginPipelineFor('runWithMdast'))
      .use(this.transformerMdastToHast)
      .use(this.createPluginPipelineFor('runWithHast'))
      .use(this.createPluginPipelineFor('runAfter'));
  }

  public run(sources: SourceConfig[]): Promise<DocfyResult> {
    return new Promise((resolve, reject) => {
      this.pipeline.run(sources, (err: unknown, ctx: Context): void => {
        if (err) {
          reject(err);
        } else {
          resolve({
            content: ctx.pages,
            staticAssets: ctx.staticAssets,
            nestedPageMetadata: transformToNestedPageMetadata(
              ctx.pages.map(p => p.meta),
              ctx.options.labels
            ),
          });
        }
      });
    });
  }

  private transformerMdastToHast(ctx: Context): void {
    ctx.pages.forEach(page => {
      const hast = ctx.rehype.runSync(page.ast, page.vFile);
      page.ast = hast;

      page.demos?.forEach(demo => {
        const hast = ctx.rehype.runSync(demo.ast, demo.vFile);
        demo.ast = hast;
      });
    });
  }

  private createPluginPipelineFor(
    funcType: 'runBefore' | 'runAfter' | 'runWithMdast' | 'runWithHast'
  ): (ctx: Context) => void {
    function isPluginWithOptionsFunction(
      plugin: Plugin | PluginWithOptions | PluginWithOptionsFunction
    ): plugin is PluginWithOptionsFunction {
      return typeof plugin === 'function' && '__isOptionsFunction' in plugin;
    }

    function isPluginWithOptions(
      plugin: Plugin | PluginWithOptions | PluginWithOptionsFunction
    ): plugin is PluginWithOptions {
      return '__options' in plugin;
    }

    const plugins = this.plugins;

    return function (ctx: Context): void {
      plugins.forEach(plugin => {
        let options: unknown;

        if (isPluginWithOptionsFunction(plugin)) {
          plugin = plugin();
        }

        if (isPluginWithOptions(plugin)) {
          options = plugin.__options;
        }

        const fn = plugin[funcType];
        if (fn) {
          fn(ctx, options as never);
        }
      });
    };
  }

  private initializePipeline(sources: SourceConfig[]): Context {
    const ctx = this.context;

    // Clear previous results to ensure fresh content on each run
    ctx.pages = [];
    ctx.staticAssets = [];

    sources.forEach(item => {
      let repoEditUrl: string | null;

      if (item.repository || ctx.options.repository) {
        repoEditUrl = getRepoEditUrl(
          item.root,
          item.repository?.url || ctx.options.repository?.url || '',
          item.repository?.editBranch || ctx.options.repository?.editBranch
        );
      }
      const files = glob.sync(item.pattern, {
        cwd: item.root,

        ignore: [...DEFAULT_IGNORE, ...(item.ignore || [])],
        absolute: true,
      });

      debug('Source Files', files);

      files.forEach(file => {
        ctx.pages.push(this.createPage(item, file, repoEditUrl));
      });
    });

    return ctx;
  }

  private createPage(
    sourceConfig: SourceConfig,
    fullPath: string,
    repoEditUrl?: string | null
  ): PageContent {
    const relativePath = fullPath.replace(path.join(sourceConfig.root, '/'), '');

    const vFile = toVfile.readSync(fullPath);
    const markdown = vFile.contents.toString();
    const parsed = this.context.remark.parse(vFile);
    const ast = this.context.remark.runSync(parsed, vFile);

    const frontmatter: Record<string, unknown> =
      (vFile.data as { frontmatter?: Record<string, unknown> }).frontmatter || {};

    let url: string;
    let title = inferTitle(ast);
    if (typeof frontmatter.title === 'string') {
      title = frontmatter.title;
    }
    // Fallback for title
    if (typeof title === 'undefined' || title === '') {
      title = path.parse(fullPath).name;
    }

    if (typeof frontmatter.url === 'string') {
      url = generateAutoUrl(frontmatter.url, sourceConfig.urlPrefix, sourceConfig.urlSuffix);
    } else if (sourceConfig.urlSchema === 'manual') {
      url = generateManualUrl(
        relativePath,
        frontmatter,
        sourceConfig.urlPrefix,
        sourceConfig.urlSuffix
      );
    } else {
      url = generateAutoUrl(relativePath, sourceConfig.urlPrefix, sourceConfig.urlSuffix);
    }

    // Add fallback order for index pages
    if (url.length > 0 && url[url.length - 1] === '/' && typeof frontmatter.order === 'undefined') {
      (frontmatter.order as undefined | number) = -1;
    }

    let editUrl = '';

    if (repoEditUrl) {
      editUrl = repoEditUrl.replace('{filepath}', relativePath);
    }

    const metadata: PageMetadata = {
      url,
      title,
      editUrl,
      relativeUrl: undefined,
      relativePath,
      headings: [],
      frontmatter: frontmatter,
      pluginData: {},
      parentLabel: undefined,
    };

    return {
      meta: metadata,
      source: relativePath,
      sourceConfig,
      vFile,
      ast,
      markdown,
      rendered: '',
      pluginData: {},
    };
  }
}

export default Docfy;
module.exports = Docfy;
