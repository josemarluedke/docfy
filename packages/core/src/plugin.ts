import { Plugin, PluginOptions, PluginWithOptions, PluginWithOptionsFunction } from './types';

function plugin(handlers: Plugin): Plugin {
  return handlers;
}

function pluginWithOptions<T = PluginOptions>(handlers: Plugin<T>): PluginWithOptionsFunction<T> {
  function optionsFunction(options?: T): PluginWithOptions<T> {
    return {
      __options: options,
      ...handlers,
    };
  }
  optionsFunction.__isOptionsFunction = true;

  return optionsFunction;
}

plugin.withOptions = pluginWithOptions;

export default plugin;
module.exports = plugin;
