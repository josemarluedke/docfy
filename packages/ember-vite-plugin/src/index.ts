import type { Plugin, ResolvedConfig } from 'vite';
import type { PluginContext } from 'rollup';
import Docfy from '@docfy/core';
import type { DocfyConfig } from '@docfy/core/lib/types';
import path from 'path';
import { loadDocfyConfig, DocfyVitePluginOptions } from './config.js';
import { createVirtualModules } from './virtual-modules.js';
import { processMarkdown } from './markdown-processor.js';
import {
  generateTemplatePath,
  generatePageTemplate
} from './gjs-generator.js';
import debugFactory from 'debug';
import fs from 'fs';
import fastGlob from 'fast-glob';

const debug = debugFactory('@docfy/ember-vite-plugin');

function findSourceConfigForFile(filePath: string, docfyInstance: Docfy): any {
  // Access the docfy config sources
  const sources = (docfyInstance as any).context?.options?.sources || [];

  for (const source of sources) {
    const sourcePath = path.resolve(source.root);
    const resolvedFilePath = path.resolve(filePath);

    if (resolvedFilePath.startsWith(sourcePath)) {
      return source;
    }
  }

  return null;
}

export default function docfyVitePlugin(
  options: DocfyVitePluginOptions = {}
): Plugin[] {
  const { root = process.cwd(), hmr = true, ...docfyOptions } = options;

  function shouldProcessFile(filePath: string): boolean {
    if (!docfyConfig?.sources) return false;

    return docfyConfig.sources.some((source) => {
      const sourceRoot = path.resolve(source.root || root);
      const resolvedFilePath = path.resolve(filePath);

      // Check if file is within source root
      if (!resolvedFilePath.startsWith(sourceRoot)) {
        return false;
      }

      // Simple pattern matching for .md files
      return filePath.endsWith('.md');
    });
  }

  async function getDocfySourceFiles(): Promise<string[]> {
    if (!docfyConfig?.sources) return [];

    const allFiles: string[] = [];

    for (const source of docfyConfig.sources) {
      const sourceRoot = path.resolve(source.root || root);
      const pattern = source.pattern || '**/*.md';
      const ignore = source.ignore || [];

      try {
        const files = await fastGlob(pattern, {
          cwd: sourceRoot,
          absolute: true,
          ignore: ['node_modules/**', ...ignore]
        });

        allFiles.push(...files);
        debug('Found source files for pattern', {
          pattern,
          sourceRoot,
          fileCount: files.length
        });
      } catch (error) {
        debug('Error globbing files', { pattern, sourceRoot, error });
      }
    }

    return allFiles;
  }

  let config: ResolvedConfig;
  let docfyConfig: DocfyConfig;
  let docfyInstance: Docfy;
  let virtualModules: ReturnType<typeof createVirtualModules>;

  return [
    {
      name: 'docfy-ember-vite-plugin:config',
      enforce: 'pre', // Ensure this runs before other plugins
      configResolved(resolvedConfig) {
        config = resolvedConfig;
        debug('Vite config resolved', {
          command: config.command,
          mode: config.mode
        });
      },

      async buildStart() {
        debug('Loading Docfy configuration...');
        docfyConfig = await loadDocfyConfig(root, docfyOptions);
        debug('Docfy configuration loaded', {
          sources: docfyConfig.sources?.length,
          sourcesDetails: docfyConfig.sources?.map((s) => ({
            root: s.root,
            pattern: s.pattern,
            urlPrefix: s.urlPrefix
          })),
          root
        });
        docfyInstance = new Docfy(docfyConfig);
        virtualModules = createVirtualModules();

        // Add all Docfy source files to Vite's watch list
        if (config.command === 'serve') {
          const sourceFiles = await getDocfySourceFiles();
          sourceFiles.forEach((file) => {
            debug('Adding file to watch list', { file });
            this.addWatchFile(file);
          });
        }

        // Process markdown files immediately to populate virtual modules
        try {
          debug('Processing markdown files...');
          const result = await docfyInstance.run(docfyConfig.sources as any);
          virtualModules.updateResult(result);
          debug('Markdown processing completed', {
            contentCount: result.content.length,
            staticAssetsCount: result.staticAssets.length
          });

          // Generate templates for both development and production using file system writes
          debug('Writing templates to file system...');
          result.content.forEach((page) => {
            const templatePath = generateTemplatePath(page.meta.url);
            const pageTemplate = generatePageTemplate(page, this);

            // Write template to file system
            const fullPath = path.join(process.cwd(), templatePath);
            const dir = path.dirname(fullPath);

            debug('Writing template to file system', {
              url: page.meta.url,
              templatePath,
              fullPath,
              command: config.command
            });

            // Ensure directory exists
            if (!fs.existsSync(dir)) {
              fs.mkdirSync(dir, { recursive: true });
            }

            // Write template file
            fs.writeFileSync(fullPath, pageTemplate);
          });
        } catch (error) {
          debug('Error processing markdown files', { error });
        }

        debug('Docfy instance created and virtual modules initialized');
      },

      resolveId(id) {
        debug('Attempting to resolve ID', { id });
        const resolved = virtualModules.resolveId(id);
        if (resolved) {
          debug('Resolved virtual module', { id, resolved });
        } else if (id.startsWith('virtual:')) {
          debug('Failed to resolve virtual module', { id });
        }
        return resolved;
      },

      load(id) {
        const loaded = virtualModules.load(id);
        if (loaded) {
          debug('Loaded virtual module', { id, contentLength: loaded.length });
        }
        return loaded;
      },

      async transform(code, id) {
        if (!shouldProcessFile(id)) return null;

        debug('Processing markdown file', { id });
        const result = await processMarkdown(code, id, docfyInstance, config);

        // For development mode, templates are generated in buildStart
        // For production, templates are generated in the second plugin's generateBundle hook
        // No need to generate templates here to avoid conflicts

        return result;
      },

      generateBundle() {
        debug('Generating bundle assets...');
        virtualModules.generateAssets(this);
      },

      ...(hmr && {
        async handleHotUpdate(ctx) {
          if (config?.command === 'serve' && shouldProcessFile(ctx.file)) {
            debug('HMR update for markdown file', { file: ctx.file });

            // Regenerate templates when markdown files change
            try {
              debug('Regenerating templates due to file change...');
              const result = await docfyInstance.run(
                docfyConfig.sources as any
              );
              virtualModules.updateResult(result);

              // Find the specific page that corresponds to the changed file
              const changedPage = result.content.find((page) => {
                // Check if this page's source file matches the changed file
                return page.vFile.path === ctx.file;
              });

              if (changedPage) {
                // Only regenerate the template for the changed file
                const templatePath = generateTemplatePath(changedPage.meta.url);
                // Create a minimal context-like object for component generation
                const contextForGeneration = {
                  emitFile: () => {}, // Not needed in HMR - we write directly to filesystem
                } as unknown as PluginContext;
                const pageTemplate = generatePageTemplate(changedPage, contextForGeneration);

                const fullPath = path.join(process.cwd(), templatePath);
                const dir = path.dirname(fullPath);

                debug('Regenerating template file for changed page', {
                  url: changedPage.meta.url,
                  templatePath,
                  fullPath,
                  changedFile: ctx.file
                });

                // Ensure directory exists
                if (!fs.existsSync(dir)) {
                  fs.mkdirSync(dir, { recursive: true });
                }

                // Write template file
                fs.writeFileSync(fullPath, pageTemplate);

                // Invalidate the specific template file in Vite's module graph
                const templateModule =
                  ctx.server.moduleGraph.getModuleById(fullPath);
                if (templateModule) {
                  ctx.server.reloadModule(templateModule);
                } else {
                  // If template module doesn't exist in graph, try to find it by looking at imports
                  const modules =
                    ctx.server.moduleGraph.getModulesByFile(fullPath);
                  if (modules) {
                    for (const mod of modules) {
                      ctx.server.reloadModule(mod);
                    }
                  }
                }
              } else {
                debug('No matching page found for changed file', {
                  file: ctx.file
                });
              }

              debug('Template regenerated successfully');
            } catch (error) {
              debug('Error regenerating templates', { error });
            }

            // Return empty array to prevent default HMR behavior
            return [];
          }
        }
      })
    },

    {
      name: 'docfy-ember-vite-plugin:generate',
      enforce: 'pre', // Ensure this runs before other plugins

      async generateBundle() {
        debug('Generating bundle assets...');
        // Generate virtual module assets, but not templates (already handled in first plugin)
        virtualModules.generateAssets(this);
      }
    }
  ];
}

// Export types for consumers
export type { DocfyConfig } from '@docfy/core/lib/types';
export type { DocfyVitePluginOptions } from './config.js';
