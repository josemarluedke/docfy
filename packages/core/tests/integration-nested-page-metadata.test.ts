import Docfy from '../src';
import { DocfyResult } from '../src/types';
import path from 'path';

const root = path.resolve(__dirname, './__fixtures__/monorepo');

describe('Generates runtime output', () => {
  let result: DocfyResult;

  beforeAll(async () => {
    const docfy = new Docfy({
      tocMaxDepth: 2,
      repository: {
        url: 'https://github.com/user/repo',
      },
    });
    result = await docfy.run([
      {
        root,
        urlPrefix: 'docs',
        urlSchema: 'manual',
        pattern: '**/*.md',
      },
    ]);
  });

  test('it should have generated the nestedPageMetadata', () => {
    expect(result.nestedPageMetadata).toMatchSnapshot();
  });
});
