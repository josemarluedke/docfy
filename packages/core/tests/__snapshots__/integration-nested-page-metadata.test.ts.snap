// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates runtime output it should have generated the nestedPageMetadata 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [],
              "label": "components",
              "name": "components",
              "pages": [
                {
                  "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/packages/package1/components/button.md",
                  "frontmatter": {
                    "category": "category1",
                    "order": 0,
                    "subcategory": "components",
                  },
                  "headings": [],
                  "parentLabel": "components",
                  "pluginData": {},
                  "relativePath": "packages/package1/components/button.md",
                  "relativeUrl": "button",
                  "title": "Button Component",
                  "url": "/docs/category1/components/button",
                },
                {
                  "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/packages/package1/components/form/index.md",
                  "frontmatter": {
                    "category": "category1",
                    "order": 1,
                    "subcategory": "components",
                  },
                  "headings": [],
                  "parentLabel": "components",
                  "pluginData": {},
                  "relativePath": "packages/package1/components/form/index.md",
                  "relativeUrl": "form",
                  "title": "Form Component",
                  "url": "/docs/category1/components/form",
                },
                {
                  "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/packages/package1/components/form/something-else.md",
                  "frontmatter": {
                    "category": "category1",
                    "order": 2,
                    "subcategory": "components",
                  },
                  "headings": [],
                  "parentLabel": "components",
                  "pluginData": {},
                  "relativePath": "packages/package1/components/form/something-else.md",
                  "relativeUrl": "something-else",
                  "title": "something-else",
                  "url": "/docs/category1/components/something-else",
                },
              ],
            },
          ],
          "label": "category1",
          "name": "category1",
          "pages": [
            {
              "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/packages/package1/docs/intro.md",
              "frontmatter": {
                "category": "category1",
              },
              "headings": [],
              "parentLabel": "category1",
              "pluginData": {},
              "relativePath": "packages/package1/docs/intro.md",
              "relativeUrl": "intro",
              "title": "Docs Intro",
              "url": "/docs/category1/intro",
            },
          ],
        },
        {
          "children": [],
          "label": "category2",
          "name": "category2",
          "pages": [
            {
              "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/packages/package2/docs/overview.md",
              "frontmatter": {
                "category": "category2",
                "order": 0,
                "title": "Overview",
              },
              "headings": [],
              "parentLabel": "category2",
              "pluginData": {},
              "relativePath": "packages/package2/docs/overview.md",
              "relativeUrl": "overview",
              "title": "Overview",
              "url": "/docs/category2/overview",
            },
            {
              "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/packages/package2/docs/styles.md",
              "frontmatter": {
                "category": "category2",
                "order": 1,
              },
              "headings": [],
              "parentLabel": "category2",
              "pluginData": {},
              "relativePath": "packages/package2/docs/styles.md",
              "relativeUrl": "styles",
              "title": "How styles work",
              "url": "/docs/category2/styles",
            },
          ],
        },
        {
          "children": [],
          "label": "test-custom",
          "name": "test-custom",
          "pages": [
            {
              "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/docs/custom-url.md",
              "frontmatter": {
                "url": "/test-custom/url",
              },
              "headings": [],
              "parentLabel": "test-custom",
              "pluginData": {},
              "relativePath": "docs/custom-url.md",
              "relativeUrl": "url",
              "title": "This a test for the ability to change the url using frontmatter",
              "url": "/docs/test-custom/url",
            },
          ],
        },
      ],
      "label": "docs",
      "name": "docs",
      "pages": [
        {
          "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/README.md",
          "frontmatter": {
            "order": -1,
          },
          "headings": [],
          "parentLabel": "docs",
          "pluginData": {},
          "relativePath": "README.md",
          "relativeUrl": "",
          "title": "The index",
          "url": "/docs/",
        },
        {
          "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/docs/how-it-works.md",
          "frontmatter": {
            "order": 0,
          },
          "headings": [],
          "parentLabel": "docs",
          "pluginData": {},
          "relativePath": "docs/how-it-works.md",
          "relativeUrl": "how-it-works",
          "title": "How it works?",
          "url": "/docs/how-it-works",
        },
        {
          "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/docs/about.md",
          "frontmatter": {
            "order": 1,
          },
          "headings": [
            {
              "depth": 2,
              "id": "history",
              "title": "History",
            },
            {
              "depth": 2,
              "id": "where-it-came-from",
              "title": "Where it came from",
            },
            {
              "depth": 2,
              "id": "another-topic",
              "title": "Another topic",
            },
            {
              "depth": 2,
              "id": "2-level",
              "title": "2 level",
            },
          ],
          "parentLabel": "docs",
          "pluginData": {},
          "relativePath": "docs/about.md",
          "relativeUrl": "about",
          "title": "About the project",
          "url": "/docs/about",
        },
        {
          "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/docs/html.md",
          "frontmatter": {},
          "headings": [],
          "parentLabel": "docs",
          "pluginData": {},
          "relativePath": "docs/html.md",
          "relativeUrl": "html",
          "title": "HTML should work",
          "url": "/docs/html",
        },
        {
          "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/docs/math.md",
          "frontmatter": {},
          "headings": [],
          "parentLabel": "docs",
          "pluginData": {},
          "relativePath": "docs/math.md",
          "relativeUrl": "math",
          "title": "Testing math remark and rehype plugins",
          "url": "/docs/math",
        },
        {
          "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/docs/static-asset.md",
          "frontmatter": {},
          "headings": [],
          "parentLabel": "docs",
          "pluginData": {},
          "relativePath": "docs/static-asset.md",
          "relativeUrl": "static-asset",
          "title": "This is a markdown that has a image",
          "url": "/docs/static-asset",
        },
        {
          "editUrl": "https://github.com/user/repo/edit/master/packages/core/tests/__fixtures__/monorepo/docs/table.md",
          "frontmatter": {},
          "headings": [],
          "parentLabel": "docs",
          "pluginData": {},
          "relativePath": "docs/table.md",
          "relativeUrl": "table",
          "title": "Table example",
          "url": "/docs/table",
        },
      ],
    },
  ],
  "label": "/",
  "name": "/",
  "pages": [],
}
`;
